README
NAME
    Function::Fallback::CoreOrPP - Functions that use non-core XS module but
    provide pure-Perl/core fallback

VERSION
    This document describes version 0.06 of Function::Fallback::CoreOrPP
    (from Perl distribution Function-Fallback-CoreOrPP), released on
    2014-09-16.

SYNOPSIS
     use Function::Fallback::CoreOrPP qw(clone unbless uniq);

     my $clone = clone({blah=>1});
     my $unblessed = unbless($blessed_ref);
     my @uniq  = uniq(1, 3, 2, 1, 4);  # -> (1, 3, 2, 4)

DESCRIPTION
    This module provides functions that use non-core XS modules (for best
    speed, reliability, feature, etc) but falls back to those that use core
    XS or pure-Perl modules when the non-core XS module is not available.

    This module helps when you want to bootstrap your Perl application with
    a portable, dependency-free Perl script. In a vanilla Perl installation
    (having only core modules), you can use App::FatPacker to include
    non-core pure-Perl dependencies to your script.

FUNCTIONS
  clone($data) => $cloned
    Try to use Data::Clone's "clone", but fall back to using Clone::PP's
    "clone".

  unbless($ref) => $unblessed_ref
    Try to use Acme::Damn's "damn" to unbless a reference but fall back to
    shallow copying.

    NOTE: "damn()" MODIFIES the original reference. (XXX in the future an
    option to clone the reference first will be provided), while shallow
    copying will return a shallow copy.

    NOTE: The shallow copy method currently only handles blessed
    {scalar,array,hash}ref as those are the most common.

  uniq(@ary) => @uniq_ary
    Try to use List::MoreUtils's "uniq", but fall back to using slower,
    pure-Perl implementation.

SEE ALSO
    Clone::Any can also uses multiple backends, but I avoid it because I
    don't think Storable's "dclone" should be used (no Regexp support out of
    the box + must use deparse to handle coderefs).

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Function-Fallback-CoreOrPP>.

SOURCE
    Source repository is at
    <https://github.com/perlancar/perl-Function-Fallback-CoreOrPP>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Function-Fallback-Cor
    eOrPP>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    perlancar <perlancar@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by perlancar@cpan.org.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Function-Fallback-CoreOrPP

0.06    2014-09-16 (PERLANCAR)

	- Move non-core XS deps from requires to recommends [RT#98886,
	  suggested by KENTNL].


0.05    2014-04-10 (SHARYANTO)

        - Add unbless().


0.04    2014-04-10 (SHARYANTO)

        - No functional changes.

        - Require the non-core XS modules as deps. We only need optional deps
          when building fatpacked app.


0.03    2014-04-10 (SHARYANTO)

        - No functional changes.

        - Rename distribution from SHARYANTO-MaybeXS to
          Function::Fallback::CoreOrPP.


0.02    2014-04-09 (SHARYANTO)

        - Replace Storable::dclone with Clone::PP::clone.


0.01    2014-04-07 (SHARYANTO)

        - First release.


lib/Function/Fallback/CoreOrPP.pm
package Function::Fallback::CoreOrPP;

use 5.010001;
use strict;
use warnings;

our $VERSION = '0.06'; # VERSION

our $USE_NONCORE_XS_FIRST = 1;

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(
                       clone
                       unbless
                       uniq
               );

sub clone {
    my $data = shift;
    goto FALLBACK unless $USE_NONCORE_XS_FIRST;
    goto FALLBACK unless eval { require Data::Clone; 1 };

  STANDARD:
    return Data::Clone::clone($data);

  FALLBACK:
    require Clone::PP;
    return Clone::PP::clone($data);
}

sub _unbless_fallback {
    my $ref = shift;

    my $r = ref($ref);
    # not a reference
    return $ref unless $r;

    # return if not a blessed ref
    my ($r2, $r3) = "$ref" =~ /(.+)=(.+?)\(/
        or return $ref;

    if ($r3 eq 'HASH') {
        return { %$ref };
    } elsif ($r3 eq 'ARRAY') {
        return [ @$ref ];
    } elsif ($r3 eq 'SCALAR') {
        return \( my $copy = ${$ref} );
    } else {
        die "Can't handle $ref";
    }
}

sub unbless {
    my $ref = shift;

    goto FALLBACK unless $USE_NONCORE_XS_FIRST;
    goto FALLBACK unless eval { require Acme::Damn; 1 };

  STANDARD:
    return Acme::Damn::damn($ref);

  FALLBACK:
    return _unbless_fallback($ref);
}

sub uniq {
    goto FALLBACK unless $USE_NONCORE_XS_FIRST;
    goto FALLBACK unless eval { require List::MoreUtils; 1 };

  STANDARD:
    return List::MoreUtils::uniq(@_);

  FALLBACK:
    my %h;
    my @res;
    for (@_) {
        push @res, $_ unless $h{$_}++;
    }
    return @res;
}

1;
#ABSTRACT: Functions that use non-core XS module but provide pure-Perl/core fallback

__END__

=pod

=encoding UTF-8

=head1 NAME

Function::Fallback::CoreOrPP - Functions that use non-core XS module but provide pure-Perl/core fallback

=head1 VERSION

This document describes version 0.06 of Function::Fallback::CoreOrPP (from Perl distribution Function-Fallback-CoreOrPP), released on 2014-09-16.

=head1 SYNOPSIS

 use Function::Fallback::CoreOrPP qw(clone unbless uniq);

 my $clone = clone({blah=>1});
 my $unblessed = unbless($blessed_ref);
 my @uniq  = uniq(1, 3, 2, 1, 4);  # -> (1, 3, 2, 4)

=head1 DESCRIPTION

This module provides functions that use non-core XS modules (for best speed,
reliability, feature, etc) but falls back to those that use core XS or pure-Perl
modules when the non-core XS module is not available.

This module helps when you want to bootstrap your Perl application with a
portable, dependency-free Perl script. In a vanilla Perl installation (having
only core modules), you can use L<App::FatPacker> to include non-core pure-Perl
dependencies to your script.

=for Pod::Coverage ^()$

=head1 FUNCTIONS

=head2 clone($data) => $cloned

Try to use L<Data::Clone>'s C<clone>, but fall back to using L<Clone::PP>'s
C<clone>.

=head2 unbless($ref) => $unblessed_ref

Try to use L<Acme::Damn>'s C<damn> to unbless a reference but fall back to
shallow copying.

NOTE: C<damn()> B<MODIFIES> the original reference. (XXX in the future an option
to clone the reference first will be provided), while shallow copying will
return a shallow copy.

NOTE: The shallow copy method currently only handles blessed
{scalar,array,hash}ref as those are the most common.

=head2 uniq(@ary) => @uniq_ary

Try to use L<List::MoreUtils>'s C<uniq>, but fall back to using slower,
pure-Perl implementation.

=head1 SEE ALSO

L<Clone::Any> can also uses multiple backends, but I avoid it because I don't
think L<Storable>'s C<dclone> should be used (no Regexp support out of the box +
must use deparse to handle coderefs).

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Function-Fallback-CoreOrPP>.

=head1 SOURCE

Source repository is at L<https://github.com/perlancar/perl-Function-Fallback-CoreOrPP>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Function-Fallback-CoreOrPP>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

perlancar <perlancar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by perlancar@cpan.org.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Functions that use non-core XS module but provide pure-Perl/core fallback",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Function-Fallback-CoreOrPP",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Function::Fallback::CoreOrPP",
  "PREREQ_PM" => {
    "Clone::PP" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Clone::PP" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



